README.md
# OpenMLDB ChatGPT Plugin

The ChatGPT plugin to enhance OpenMLDB. Here are the features:

* Integrate **GPT models** with OpenMLDB service
* Automatically **generate SQL** and **correct SQL**
* Enable to update database with **chat interface**
* New using experience of database with **AI prompts**

## Install

Install with [python package](https://pypi.org/project/openmldb-chatgpt/).

```
pip install openmldb-chatgpt
```

Use pre-built docker image (OpenMLDB included).

```
docker run -it tobegit3hub/openmldb-chatgpt
```

## Usage

Install [OpenMLDB](https://github.com/4paradigm/OpenMLDB) with [PortableOpenMLDB](https://github.com/tobegit3hub/PortableOpenMLDB). No need to install for docker container.

Register [OpenAI account](https://openai.com) and get the API key. Make sure you can access the OpenAI API.

Run `openmldb-chatgpt` to connect with OpenMLDB and use OpenAI API key.

![](./images/openmldb_chatgpt_cli.png)

## Prompts

For beginners, AI model can teach you how to use OpenMLDB and database.

![](images/prompt1.png)

![](images/prompt1-1.png)

All the SQL will be executed in OpenMLDB and explained by AI model.

![](images/prompt2.png)

![](images/prompt2-1.png)

![](images/prompt2-2.png)

If you create a little more complicated table, AI model will try to understand the meaning with your input. 

![](images/prompt3.png)

For senior SQL users, AI model can help to debug and correct your SQL.

![](images/prompt4.png)

AI model can help to analyse the complex SQL and explain it in natural language as well.

![](images/prompt5.png)

You can ask AI model to show more info about OpenMLDB cluster without writing specified SQL.

![](images/prompt6.png)

**Please try by yourself and discover more magic functions with the powerful AI model and OpenMLDB!** 


openmldb_chatgpt/history.txt

# 2023-03-16 15:44:00.026686
+cherry

# 2023-03-16 15:45:29.446453
+q

# 2023-03-16 15:45:35.415170
+SELECT

# 2023-03-16 15:46:51.193367
+Explain the following sql: "SELECT * FROM t1" 

# 2023-03-16 15:47:44.616850
+q

# 2023-03-16 15:47:48.573115
+sdf

# 2023-03-16 15:48:04.167470
+what is your name

# 2023-03-16 15:49:39.197514
+what is your name hallo

# 2023-03-16 15:50:01.139162
+event

# 2023-03-16 15:50:31.324904
+how old are you

# 2023-03-16 15:51:49.078040
+q

# 2023-03-16 15:52:01.632307
+tell me how to create database in SQ

# 2023-03-16 15:54:30.007291
+q

# 2023-03-16 15:54:34.987549
+sdfasfasf

# 2023-03-16 15:58:10.304430
+q

# 2023-03-16 15:59:54.076661
+SELECT and fail

# 2023-03-16 16:00:52.745640
+q

# 2023-03-16 16:01:04.405887
+quit

# 2023-03-16 16:01:22.654945
+q

# 2023-03-16 16:37:33.871097
+how to make money with chatgpt

# 2023-03-16 17:21:30.251229
+q

# 2023-03-16 17:21:50.670818
+how to make money with chatgpt

# 2023-03-16 17:21:56.299518
+continues

# 2023-03-16 17:22:05.286208
+go on

# 2023-03-16 17:22:31.651858
+show me a SQL to create table

# 2023-03-16 17:22:40.644334
+go on

# 2023-03-16 17:22:59.039236
+q

# 2023-03-16 17:23:33.053472
+SELECT and fail

# 2023-03-16 17:24:26.039607
+tell me how to create database in SQ

# 2023-03-16 17:24:33.708582
+go on

# 2023-03-16 17:24:36.794407
+goon

# 2023-03-16 17:24:56.713741
+give me the above sql without comments

# 2023-03-16 17:25:05.599354
+go

# 2023-03-16 17:25:10.428974
+goon

# 2023-03-16 17:25:19.376398
+continues

# 2023-03-16 17:25:44.816951
+if a = 1, b=3, c=5, what above d

# 2023-03-16 17:26:08.561774
+if a is 1, b is3, c is 5, d is ?

# 2023-03-16 17:26:21.797287
+if a is 1, b is 2, c is 3, d is ?

# 2023-03-16 17:26:26.459591
+continues

# 2023-03-16 17:26:56.498415
+q

# 2023-03-16 17:27:06.530028
+if a is 1, b is 2, c is 3, d is ?

# 2023-03-16 17:27:18.030903
+how about e

# 2023-03-16 17:27:31.496697
+tell me a joke in Chinese

# 2023-03-16 17:27:37.519410
+continues

# 2023-03-16 17:28:04.210293
+stop

# 2023-03-16 17:28:16.515528
+what is your name

# 2023-03-16 17:28:30.586359
+translate hala into Japanese

# 2023-03-16 17:28:38.737025
+continues

# 2023-03-16 17:28:48.746793
+q

# 2023-03-16 18:23:42.854038
+next

# 2023-03-16 18:23:47.654780
+q

# 2023-03-16 18:23:59.151375
+ORDER BY  

# 2023-03-16 18:24:51.500072
+q

# 2023-03-16 18:25:07.538644
+sdfa INSERT s 

# 2023-03-16 18:25:28.504170
+q

# 2023-03-16 18:27:49.899689
+sdfaf sdfasfasf afdsa SELECT

# 2023-03-16 18:28:34.107292
+q

# 2023-03-16 18:31:43.020791
+hellow man

# 2023-03-16 18:33:08.937395
+q

# 2023-03-16 18:33:30.554133
+sdfa

# 2023-03-16 18:34:24.516061
+q

# 2023-03-16 18:34:40.779626
+hsll

# 2023-03-16 18:35:42.021628
+q

# 2023-03-16 18:35:47.086406
+whl

# 2023-03-16 18:36:20.888877
+hello

# 2023-03-16 18:59:09.802046
+q

# 2023-03-16 19:05:33.387468
+show jobs;

# 2023-03-16 19:05:36.289436
+show jobss;

# 2023-03-16 19:07:05.986273
+q

# 2023-03-16 19:07:09.701504
+ehll

# 2023-03-16 19:07:36.201138
+sd

# 2023-03-16 19:08:02.229104
+show josdf;

# 2023-03-16 19:08:17.855191
+sdf;

# 2023-03-16 19:09:35.141714
+q

# 2023-03-16 19:09:38.557676
+dsf

# 2023-03-16 19:10:16.536329
+show jobs;

# 2023-03-16 19:10:23.196172
+select 100;

# 2023-03-16 19:12:41.650694
+q

# 2023-03-16 19:12:46.329396
+select 100;

# 2023-03-16 19:19:43.659904
+SELECT 100;

# 2023-03-16 19:19:47.120735
+show jobs;

# 2023-03-16 19:19:57.665008
+create database db2;

# 2023-03-16 19:22:47.484725
+q

# 2023-03-16 19:22:51.918437
+show jobs;

# 2023-03-16 19:23:02.311443
+select 1;

# 2023-03-16 19:26:56.129489
+q

# 2023-03-16 19:27:00.674228
+show jobs;

# 2023-03-16 19:27:18.692058
+select 1;

# 2023-03-16 19:27:32.472339
+create database db3;

# 2023-03-16 19:37:51.685229
+select 1;

# 2023-03-16 19:37:54.367313
+show jobs;

# 2023-03-16 19:38:49.962634
+q

# 2023-03-16 19:39:08.757258
+show jobs;

# 2023-03-16 19:39:38.169887
+create database db3;

# 2023-03-16 19:40:07.088964
+create database db4

# 2023-03-16 19:40:12.963893
+show databases;

# 2023-03-16 19:40:16.793879
+show tables;

# 2023-03-16 19:40:23.173967
+use db1;

# 2023-03-16 19:40:34.249807
+use db10;

# 2023-03-16 19:40:37.855890
+show tables;

# 2023-03-16 19:40:45.243052
+create table t1 (col1 int);

# 2023-03-16 19:40:49.598134
+show tables;

# 2023-03-16 19:41:13.387192
+select * from t1;

# 2023-03-16 19:42:35.569534
+q

# 2023-03-16 19:42:39.560088
+show jobs;

# 2023-03-16 19:42:44.047332
+select 1;

# 2023-03-16 19:42:49.133453
+use db10;

# 2023-03-16 19:42:52.054037
+select * from t1;

# 2023-03-16 19:43:25.233589
+SET @@execute_mode="offline";

# 2023-03-16 19:43:31.454546
+select 10;

# 2023-03-16 19:43:34.227346
+show jobs;

# 2023-03-16 19:43:35.881244
+show job 1;

# 2023-03-16 19:43:45.431946
+show joblog 1;

# 2023-03-16 19:45:35.485389
+q

# 2023-03-16 19:45:38.456707
+show jobs;

# 2023-03-16 19:45:41.541463
+show joblog 1;

# 2023-03-16 19:46:02.201886
+q

# 2023-03-16 19:46:07.640104
+show joblog 1;

# 2023-03-16 19:46:24.000747
+q

# 2023-03-16 19:58:20.361633
+help

# 2023-03-16 19:58:32.697764
+q

# 2023-03-16 20:08:02.123744
+help

# 2023-03-16 20:08:13.482351
+SELECT 100;

# 2023-03-16 20:09:02.517243
+q

# 2023-03-16 20:14:28.159227
+帮我生成一个创建数据库的SQL

# 2023-03-16 20:16:28.976083
+q

# 2023-03-16 20:16:39.812100
+帮我生成一个创建数据库的SQL

# 2023-03-16 20:17:05.814438
+帮我生成一个名为db1的数据库的SQL

# 2023-03-16 20:17:12.517985
+go on

# 2023-03-16 20:17:41.225058
+q

# 2023-03-16 20:18:36.072003
+生成创建db1数据库的SQL

# 2023-03-16 20:19:06.226292
+生成创建名为db1的数据库的SQL

# 2023-03-16 20:19:28.456434
+执行上面的SQL语句

# 2023-03-16 20:20:53.872953
+go on

# 2023-03-16 20:21:40.332597
+帮我查询系统有多少张表

# 2023-03-16 20:22:28.833551
+q

# 2023-03-16 20:22:35.157137
+ 帮我查询系统有多少张表

# 2023-03-16 20:22:57.729613
+show jobs;

# 2023-03-16 20:37:34.557413
+help

# 2023-03-16 20:38:31.136687
+show jobs;

# 2023-03-16 20:38:35.166370
+select 1234;

# 2023-03-16 20:38:44.593427

openmldb_chatgpt/print_util.py
# Copyright 2023
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

class PrintUtil:

    @staticmethod
    def print_colorful(color, text, end='\n'):
        color_codes = {
            'black': 30,
            'red': 31,
            'green': 32,
            'yellow': 33,
            'blue': 34,
            'magenta': 35,
            'cyan': 36,
            'white': 37,
        }
        color_sequence = f"\033[{color_codes[color]}m"
        reset_sequence = "\033[0m"
        print(f"{color_sequence}{text}{reset_sequence}", end=end)

    @staticmethod
    def user_print(text, end='\n'):
        PrintUtil.print_colorful("blue", "USER > " + text, end)

    @staticmethod
    def gpt_print(text, end='\n'):
        PrintUtil.print_colorful("green", "GPT > " + text, end)

    @staticmethod
    def openmldb_print(text, end='\n'):
        PrintUtil.print_colorful("yellow", "OpenMLDB > " + text, end)
